### PR TITLE
Issue 211

### PR DESCRIPTION
**Pull request by sharkovadarya**
Wrong completion list after rules./checkpoints. https://github.com/JetBrains-Research/snakecharm/issues/211

### CODE REVIEW

File: src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt [hunk 1]
```diff
@@ -87,9 +87,17 @@ private class SmkRulesAndCheckpointsObjectsCompletionProvider : CompletionProvid
     ) {
         val variants = collectVariantsForElement(parameters.position)
 
+        val parentElement = parameters.position.parent
+        val referencedObject = parameters.withPosition(parentElement, parentElement.textOffset).originalPosition?.parent
+        if (referencedObject is PyReferenceExpression) {
+            when (referencedObject.name) {
+                SnakemakeNames.SMK_VARS_RULES -> variants.removeAll { it.second is SmkCheckPoint }
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt]: Not clear, why we collect both rules and checkpoints from an index in line 88 and then remove rules or checkpoints here. Why not to collect the only required type of elements from the index? Index keys differ for rules and checkpoints.

Comment (sharkovadarya) [reply src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt]: done

File: src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt [hunk 2]
```diff
@@ -132,34 +151,24 @@ private class SmkRulesAndCheckpointsObjectsCompletionProvider : CompletionProvid
 
 // the following functions can be used to implement completion for any section/object referring to rule names
 
-private fun collectVariantsForElement(element: PsiElement): MutableList<Pair<String, SmkRuleOrCheckpoint>> {
+private fun <Psi: SmkRuleOrCheckpoint> collectVariantsForElement(
+        element: PsiElement,
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt]: `collectVariantsForElement` is private and is used only here, but has:  too generic and not type-safe signature, e.g. you could call this method with use `indexKey` for rules and pass checkpoints section class. Also its usages introduce code duplication.

Let's use smth simpler, e.g. `collectRuleorCheckpointVariants(element, isCheckPoint=True): List<Pair<String, SmkRuleOrCheckpoint>>`

File: src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt [hunk 3]
```diff
@@ -132,34 +151,24 @@ private class SmkRulesAndCheckpointsObjectsCompletionProvider : CompletionProvid
...
+        indexKey: StubIndexKey<String, Psi>,
+        sectionClass: Class<Psi>,
+        currentFileDeclarationsFunction: (SmkFile) -> List<Pair<String, Psi>>
+): MutableList<Pair<String, SmkRuleOrCheckpoint>> {
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt]: Not sure, that usage of `MutableList` here provides noticeable benefits. Immutable collections are more  readable / easy to understand immutable collections. This method force use to think who and why modifies return result of this method, track usages etc.

File: src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt [hunk 4]
```diff
@@ -52,7 +53,18 @@ private class SmkRuleNameReferenceCompletionProvider : CompletionProvider<Comple
             context: ProcessingContext,
             result: CompletionResultSet
     ) {
-        val variants = collectVariantsForElement(parameters.position)
+        val variants = collectVariantsForElement(
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt]: Class names for reference providers don't match with implementation, e.g.:
* `SmkRuleNameReferenceCompletionProvider` - adds both rule and checkpoint names into completion
* `SmkRulesAndCheckpointsObjectsCompletionProvider` - adds rule or checkpoint, but not both

File: src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt (hunk 2 above)
Comment (sharkovadarya) [reply src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt]: done

File: src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt (hunk 3 above)
Comment (sharkovadarya) [reply src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt]: done

File: src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt (hunk 4 above)
Comment (sharkovadarya) [reply src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt]: done

File: src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt [hunk 5]
```diff
@@ -26,18 +25,18 @@ class SmkRulesAndCheckpointsCompletionContributor : CompletionContributor() {
...
         )
         extend(
                 CompletionType.BASIC,
-                SmkRulesAndCheckpointsObjectsCompletionProvider.IN_SMK_RULES_OR_CHECKPOINTS_OBJECT,
-                SmkRulesAndCheckpointsObjectsCompletionProvider()
+                SmkRulesOrCheckpointsObjectsCompletionProvider.IN_SMK_RULES_OR_CHECKPOINTS_OBJECT,
+                SmkRulesOrCheckpointsObjectsCompletionProvider()
         )
     }
 }
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt]: Here filler name better describes the provider idea. You have 2 providers, one provides completion in `localrules` and `ruleorder` sections, the other one after `rules` and `checkpoints` objects. This information in provider name is easier to read/understand compared to `rules or checkpoints provider` and `rule and checkpoint provider` which look very similar. (no action requied, i've renamed)

File: src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt [hunk 6]
```diff
@@ -151,31 +139,38 @@ private class SmkRulesAndCheckpointsObjectsCompletionProvider : CompletionProvid
...
                 .filter { (it as SmkRuleOrCheckpoint).name != null }
                 .map { (it as SmkRuleOrCheckpoint).name!! to it }
         )
     } else {
-        variants.addAll(currentFileDeclarationsFunction.invoke(element.containingFile.originalFile as SmkFile))
+        if (isCheckPoint) {
+            variants.addAll((element.containingFile.originalFile as SmkFile).collectCheckPoints())
+        } else {
+            variants.addAll((element.containingFile.originalFile as SmkFile).collectRules())
+        }
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt]: code duplication here, substing `variants.addAll((element.containingFile.originalFile as SmkFile).`...`)` is quite long and same in lines 163 and 165. (no action required)

**Pull request merged**